Create table_to_annotations.py

Bankso opened the pull request:
Addresses https://github.com/mc2-center/data-models/issues/153

This script uses file syn ids from a Dataset to identify relevant metadata records in tables (Dataset and Table syn ids are provided at runtime), then applies the metadata records to files as Synapse Annotations.

Example: `BiospecimenSpecies` and `BiospecimenFixative` annotations applied to a file. Neither of these annotations are present in file-based metadata, but were applied using this script.

<img width="300" alt="Screenshot 2025-04-01 at 5 19 42 PM" src="https://github.com/user-attachments/assets/430a90d2-a930-491b-8d27-4eb1126e972a" />

Bankso committed:
Create table_to_annotations.py

Bankso committed:
Update table_to_annotations.py

Bankso committed:
Update table_to_annotations.py

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+    
+    query = f"SELECT id FROM {dataset_id}"
+    file_table = syn.tableQuery(query).asDataFrame().fillna("")
+    files = list(file_table["id"])
```
Just FYI - pandas has a method for this conversion too!

```python
file_table["id"].tolist()
```

As far as I can tell, there's not a big difference between these two approaches, but `tolist()` can be handy whenever you need to do some pandas chaining.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+    annotations_table = syn.tableQuery(query).asDataFrame().fillna("")
+    print(f"Annotations acquired from Synapse table {source_id}")
+
+    return annotations_table
```
I think you could potentially generalize this function and combine it with the one above, maybe something like this?

```python
def get_table(syn, source_id: str, cols: str = "*") -> pd.DataFrame:
    query = f"SELECT {cols} FROM {source_id}"
    table = syn.tableQuery(query).asDataFrame().fillna("")
    print(f"Data acquired from Synapse table {source_id}"
    return table

...
...

annotations_table = get_table(...)
file_ids = get_table(..., cols="id")["id"].tolist()
```

Bankso replied:
Nice! I went ahead and added this, thank you!

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+    count = 0
+    for _, row in data_table.iterrows():
+        id = row["Biospecimen_id"]
+        if id in file_tuples.keys():
```
nit: the default behavior for iterating and checking membership in a dictionary is to look at the keys, so there's no need to include `.keys()` here!  That is, this would achieve the same thing:

```suggestion
        if id in file_tuples:
```

^^ this is a bit more Pythonic as well, since it doesn't create a new list like `keys()` would.

Bankso replied:
That makes sense!

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+    count = 0
+    for _, row in data_table.iterrows():
+        id = row[f"{component}_id"]
+        if id in tuple_dict.keys():
```
Same comment as above.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+        if id in tuple_dict.keys():
+            file_id = tuple_dict[id]
+            annotations = list(zip(column_list, list(row)))
+            annotations = apply_annotations_to_entity(syn, component, file_id, annotations)
```
Stale variable, as it's not used after this. 

Additionally, since `apply_annotations_to_entity()` doesn't return anything, assigning its result to a variable doesn't really serve a purpose and so, could be removed:

```suggestion
            apply_annotations_to_entity(syn, component, file_id, annotations)
```

Bankso replied:
Got it! I found another stale variable in `apply_annotations_to_entity()` that I went ahead and removed, too.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+
+    syn = (
+        synapseclient.login()
+    ) 
```
nit: line of code is short, so no need to break it over multiple lines.

```suggestion
    syn = synapseclient.login()
```

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,277 @@
+                           "Biospecimen Treatment Response",
+                           "Biospecimen Last Known Disease Status",
+                           "Biospecimen BioSample Identifier",
+                           "Biospecimen Description"]
```
How long did it take you to write these out?! 😆

Bankso replied:
Not long! I copied the column names from a manifest template google sheet, surrounded each name with double quotes, and joined them to a comma-separated list 😄

vpchung reviewed (changes requested):
Looks good overall! I have a couple of (nit) suggestions.

Bankso committed:
Update table_to_annotations.py

use pandas tolist() for file syn ids

Bankso committed:
Update table_to_annotations.py

Add requested changes
Add line to get_table which converts a list of columns to a quoted string for table query

Bankso commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,265 @@
+def get_table(syn, source_id: str, cols: str = "*") -> pd.DataFrame:
+    """Collect columns from a Synapse table entity and return as a Dataframe."""
+    
+    if type(cols) == list:
+        cols = ", ".join(["".join(['"', col, '"']) for col in cols])
```
I'm not sure if there's a better way to do this, but I added this to convert col lists to a string with each column name in double quotes, separated by commas (since that format is required for the table query)

vpchung replied:
That's a great idea - love it!  Since `cols` could also be a list, it would be worth to update the type-hint on L51 from `str` to `str | list`.

As for the query itself, you don't actually need the double-quotes when running it via synapseclient.  This would work too:

```python
>>> query1
'SELECT "id", "createdOn" FROM syn53293336'
>>> query2
'SELECT id, createdOn FROM syn53293336'
>>> syn.tableQuery(query1).asDataFrame()
                                                      id      createdOn
9743445_35_397eb7f4-47a9-41ed-b464-f6b4da1e0884  9743445  1711523423943
>>> syn.tableQuery(query2).asDataFrame()
                                                      id      createdOn
9743445_35_397eb7f4-47a9-41ed-b464-f6b4da1e0884  9743445  1711523423943
```

Bankso replied:
> As for the query itself, you don't actually need the double-quotes when running it via synapseclient.

That is true for column names that don't have spaces, but most of the column names we use have spaces (e.g., Biospecimen Primary Diagnosis), which (as far as I know) do need to be surrounded by double quotes

vpchung replied:
Ah yes, you're right!  Good catch.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -0,0 +1,265 @@
+    )
+    return parser.parse_args()
+
+def get_table(syn, source_id: str, cols: str = "*") -> pd.DataFrame:
```
```suggestion
def get_table(syn, source_id: str, cols: str | list = "*") -> pd.DataFrame:
```

vpchung reviewed (approved):
Other than the type-hint, LGTM! 🔥

Bankso committed:
Update utils/table_to_annotations.py

Co-authored-by: Verena Chung <9377970+vpchung@users.noreply.github.com>
